core/controllers/cron_test.py

```diff
@@ -0,0 +1,360 @@
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tests for the cron jobs."""
+
+import logging
+
+from core import jobs
+from core.controllers import cron
+from core.domain import cron_services
+from core.domain import email_manager
+from core.domain import exp_services
+from core.domain import state_domain
+from core.domain import suggestion_services
+from core.platform import models
+from core.platform.taskqueue import gae_taskqueue_services as taskqueue_services
+from core.tests import test_utils
+import feconf
+import main_cron
+
+from mapreduce import model as mapreduce_model
+import webtest
+
+(job_models, suggestion_models,) = models.Registry.import_models(
+    [models.NAMES.job, models.NAMES.suggestion])
+
+
+class SampleMapReduceJobManager(jobs.BaseMapReduceJobManager):
+    """Test job that maps over the general suggestion model."""
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [suggestion_models.GeneralSuggestionModel]
+
+
+class JobStatusMailerHandlerTests(test_utils.GenericTestBase):
+
+    def setUp(self):
+        super(JobStatusMailerHandlerTests, self).setUp()
+        self.signup(self.ADMIN_EMAIL, self.ADMIN_USERNAME)
+        self.admin_id = self.get_user_id_from_email(self.ADMIN_EMAIL)
+        self.set_admins([self.ADMIN_USERNAME])
+        self.testapp_swap = self.swap(
+            self, 'testapp', webtest.TestApp(main_cron.app))
+
+    def test_send_mail_to_admin_on_job_success(self):
+        self.login(self.ADMIN_EMAIL, is_super_admin=True)
+
+        email_subjects = []
+        email_bodies = []
+
+        def _mock_send_mail_to_admin(email_subject, email_body):
+            """Mocks email_manager.send_mail_to_admin() as its not possible to
+            send mail with self.testapp_swap, i.e with the URLs defined in
+            main_cron.
+            """
+            email_subjects.append(email_subject)
+            email_bodies.append(email_body)
+
+        send_mail_to_admin_swap = self.swap(
+            email_manager, 'send_mail_to_admin', _mock_send_mail_to_admin)
+
+        with self.testapp_swap, send_mail_to_admin_swap:
+            self.get_html_response('/cron/mail/admin/job_status')
+
+        self.assertEqual(email_subjects, ['MapReduce status report'])
+        self.assertEqual(email_bodies, ['All MapReduce jobs are running fine.'])
+
+        self.logout()
+
+    def test_send_mail_to_admin_on_job_failure(self):
+        self.login(self.ADMIN_EMAIL, is_super_admin=True)
+
+        email_subjects = []
+        email_bodies = []
+
+        def _mock_send_mail_to_admin(email_subject, email_body):
```

Comment:
Do not duplicate code. Define this as a property of the class instead, and reuse it where necessary.

As a general rule, always try to find ways to avoid duplication.

Reply:
Done